### PR TITLE
Support for proxy node in GKE tests

### PR DESCRIPTION
**Pull request by jkaniuk**
Support for proxy node (bastion) in GKE tests.
Needed for correctness tests, as some of them require SSH access to nodes.

/cc @mm4tt 
/cc @wojtek-t

### CODE REVIEW

**Comment by jkaniuk** [comment]
Will do manual tests
/hold

File: kubetest/gke.go [hunk 1]
```diff
@@ -487,15 +489,127 @@ func (g *gkeDeployer) TestSetup() error {
...
+	if err := g.setupBastion(); err != nil {
+		return err
+	}
 	if err := g.setupEnv(); err != nil {
 		return err
 	}
-	if g.sshProxyInstanceName != "" {
-		if err := setKubeShhBastionEnv(g.project, g.zone, g.sshProxyInstanceName); err != nil {
-			return err
+	g.setup = true
```
Comment (mm4tt) [inline kubetest/gke.go:498]: Why are we setting this now and weren't setting it before?

File: kubetest/gke.go [hunk 2]
```diff
@@ -487,15 +489,127 @@ func (g *gkeDeployer) TestSetup() error {
...
+	return nil
+}
+
+func (g *gkeDeployer) setupBastion() error {
+	if g.sshProxyInstanceName == "" {
+		return nil
+	}
+	var filtersToTry []string
+	// Use exact name first, VM do not have to belong to the cluster
```
Comment (mm4tt) [inline kubetest/gke.go:507]: NIT: does not

File: kubetest/gke.go [hunk 3]
```diff
@@ -102,6 +102,8 @@ type gkeDeployer struct {
 	setup          bool
 	kubecfg        string
 	instanceGroups []*ig
+	bastionIPName  string
+	zoneRegion     string
```
Comment (mm4tt) [inline kubetest/gke.go:106]: What is zoneRegion?

File: kubetest/gke.go [hunk 4]
```diff
@@ -102,6 +102,8 @@ type gkeDeployer struct {
 	setup          bool
 	kubecfg        string
 	instanceGroups []*ig
+	bastionIPName  string
```
Comment (mm4tt) [inline kubetest/gke.go:105]: What is bastionIPName? What is IPName?

File: kubetest/gke.go [hunk 5]
```diff
@@ -487,15 +489,127 @@ func (g *gkeDeployer) TestSetup() error {
...
+
+func (g *gkeDeployer) cleanupBastion() error {
+	if g.bastionIPName == "" {
+		return nil
+	}
+	region, err := g.getRegion()
+	if err != nil {
+		return fmt.Errorf("error finding region for proxy instance address: %v", err)
+	}
+	if err := control.FinishRunning(exec.Command("gcloud", "compute", "addresses", "delete", g.bastionIPName,
```
Comment (mm4tt) [inline kubetest/gke.go:607]: If the bastion VM had public IP then we won't have the address and the deletion will fail, right?

File: kubetest/gke.go [hunk 6]
```diff
@@ -487,15 +489,127 @@ func (g *gkeDeployer) TestSetup() error {
...
+	err := setKubeShhBastionEnv(g.project, zone, bastion)
+	if err == nil {
+		// Success, finishing
+		return nil
+	}
+	if !strings.Contains(err.Error(), "doesn't have an external IP address") {
+		return fmt.Errorf("setting KUBE_SSH_BASTION variable failed: %s", util.ExecError(err))
+	}
+
+	log.Printf("Trying to reserve external IP address for proxy instance")
```
Comment (mm4tt) [inline kubetest/gke.go:556]: Can we extract the code from here to a seperate method? E.g. createPublicIpForBastion or something like that

File: kubetest/gke.go (hunk 1 above)
Comment (jkaniuk) [reply kubetest/gke.go:497]: just GUI diff error, it was there before as well

File: kubetest/gke.go (hunk 2 above)
Comment (jkaniuk) [reply kubetest/gke.go]: Done

File: kubetest/gke.go (hunk 3 above)
Comment (jkaniuk) [reply kubetest/gke.go]: changed to regionCache - used in getRegion()

File: kubetest/gke.go (hunk 4 above)
Comment (jkaniuk) [reply kubetest/gke.go]: I'm sorry, it was old version of the code (with static IPs). Somehow my next push landed in the wrong place.

File: kubetest/gke.go (hunk 5 above)
Comment (jkaniuk) [reply kubetest/gke.go]: I'm sorry, it was old version of the code (with static IPs). Somehow my next push landed in the wrong place.
But address deletion was guarded by new address name, not by bastion name.

File: kubetest/gke.go (hunk 6 above)
Comment (jkaniuk) [reply kubetest/gke.go]: sorry, outdated

File: kubetest/gke.go [hunk 7]
```diff
@@ -487,15 +488,70 @@ func (g *gkeDeployer) TestSetup() error {
...
+	log.Printf("Adding NAT access config if not present")
+	control.NoOutput(exec.Command("gcloud", "compute", "instances", "add-access-config", bastion,
+		"--zone="+zone,
+		"--project="+g.project))
+
+	err := setKubeShhBastionEnv(g.project, zone, bastion)
+	if err != nil {
+		return fmt.Errorf("setting KUBE_SSH_BASTION variable failed: %s", util.ExecError(err))
 	}
-	g.setup = true
```
Comment (jkaniuk) [inline kubetest/gke.go:498]: @mm4tt : Here it is

**Comment by jkaniuk** [comment]
After first few tests looks good
/unhold

File: kubetest/gke.go [hunk 8]
```diff
@@ -692,18 +748,22 @@ func (g *gkeDeployer) ensureNat() error {
 	return nil
 }
 
-func (g *gkeDeployer) getRegion(region, zone string) (string, error) {
-	if region != "" {
-		return region, nil
+func (g *gkeDeployer) getRegion() (string, error) {
```
Comment (wojtek-t) [inline kubetest/gke.go:751]: Why do we need those changes?

Those doesn't seem to be related to the bastion changes. And if so, I would prefer them to be moved out of this PR to make is as small as possible.

Comment (jkaniuk) [reply kubetest/gke.go]: Right, after ditching static address handling, getRegion is not used anymore.
Cleaned up.

File: kubetest/gke.go (hunk 3 above)
Comment (jkaniuk) [reply kubetest/gke.go]: removed after refactoring

**Comment by wojtek-t** [comment]
/lgtm
/approve

Thanks!
